pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+      }
+      return false;
+    }
+
+    public boolean isFinal() {
+      if (this.equals(ERROR) || this.equals(COMMITTED) || this.equals(RETAINED) || this.equals(DISCARDED)) {
+        return true;
+      }
+      return false;
+    }
+  }
   private static final Logger LOGGER = LoggerFactory.getLogger(LLRealtimeSegmentDataManager.class);
   private static final int KAFKA_MAX_FETCH_TIME_MILLIS = 1000;
 
   private final LLCRealtimeSegmentZKMetadata _segmentZKMetadata;
   private final AbstractTableConfig _tableConfig;
-  private final InstanceZKMetadata _instanceZKMetadata;
   private final RealtimeTableDataManager _realtimeTableDataManager;
-  private final String _absolutePath;
-  private final ReadMode _readMode;
+  private final String _decoderClassName;
+  private final KafkaMessageDecoder _messageDecoder;
+  private final int _segmentMaxRowCount;
+  private final String _resourceDataDir;
   private final Schema _schema;
   private final ServerMetrics _serverMetrics;
   private final RealtimeSegmentImpl _realtimeSegment;
-  private volatile boolean _stopConsuming;
-  private Thread _indexingThread;
+  private long _endOffset = Long.MAX_VALUE; // No upper limit on Kafka offset
+  private long _currentOffset;
+  private volatile State _state;
+  private int _numRowsConsumed = 0;
+  private long _startOffset = 0;
+  private long _startTimeMs = 0;
+  private final String _segmentNameStr;
+  private final SegmentVersion _segmentVersion;
+
+  // Segment end criteria
+  private long _consumeEndTime = 0;
+  private volatile long _finalOffset = -1;
+  private boolean _receivedStop = false;
+
+  // It takes 30s to locate controller leader, and more if there are multiple controller failures.
+  // For now, we let 31s pass for this state transition.
+  private final int _maxTimeForConsumingToOnlineSec = 31;
+
+  private Thread _consumerThread;
+  private final String _kafkaTopic;
+  private final int _kafkaPartitionId;
+  final String _clientId;
+  private final LLCSegmentName _segmentName;
+  private final PlainFieldExtractor _fieldExtractor;
+  private final SimpleConsumerWrapper _consumerWrapper;
+  private final File _resourceTmpDir;
+  private final String _tableName;
+  private final List<String> _invertedIndexColumns;
+  private final String _sortedColumn;
+  private Logger segmentLogger = LOGGER;
+  private final String _tableStreamName;
+  private AtomicLong _lastUpdatedRawDocuments = new AtomicLong(0);
+  private final String _instance;
+  private final ServerSegmentCompletionProtocolHandler _protocolHandler;
+
+
+  private boolean endCriteriaReached() {
+    Preconditions.checkState(_state.shouldConsume(), "Incorrect state %s", _state);
+    if (_state.equals(State.INITIAL_CONSUMING)) {
+      if (_consumeEndTime > 0) {
+        // We have a time-based constraint.
+        long now = now();
+        if (now >= _consumeEndTime) {
+          segmentLogger.info("Stopping consumption due to time limit start={} now={}", _startTimeMs, now);
```

Comment:
We should display the number of rows consumed when we hit the time limit. Also, if we consumed 0 rows and try to create a segment, I think the segment creator throws an exception. In this case, it would probably be good to not close the segment and just keep consuming, it's somewhat useless to write a zero row segment.